### ISSUE TITLE
Proposal: Repository mirroring improvements

### ISSUE THREAD

**Issue by tsenart** [sourcegraph/sourcegraph#1467]
## Background

As it's implemented today, repository mirroring isn't meeting user's expectations in a number of scenarios. Whenever admins or users of a configured external service do any of the below listed actions, Sourcegraph should gracefully converge the set of repositories it mirrors to what users expect following those actions, and do so with clear representation in its UI.

This is a sub-issue of #914.

## Proposal

1. **Removed permissions**: Removing permissions from an access token used to read previously mirrored repositories from an external service **SHOULD** result in those now inaccessible repositories to be soft-deleted from Sourcegraph, with the reason for which they were deleted made clear to the user.
1. **Added permissions**: Adding extended permissions to an access token to read previously inaccessible repositories of an external service **SHOULD** result in those now accessible repositories to be automatically mirrored by Sourcegraph given they match the external service's configuration of which repositories to mirror (e.g via `github.repositoryQuery`)
1. **Renaming repositories**: Renaming a mirrored repository in an external service **SHOULD** result in a graceful repository rename operation in Sourcegraph (i.e. all data is preserved, only the name changes).
1. **Deleting repositories**: Deleting a mirrored repository in an external service **SHOULD** result in the repository being soft-deleted via a `deleted_at` timestamp in a database column and associated application logic. Soft-deleted repositories can be restored gracefully by clearing out their `deleted_at` column. Associated data like code discussions is preserved and restorable. Cached or derived data like git-server storage and search indexes is not preserved because it's cheap to re-fetch / re-compute.
1. **Modifying repositories**: Modifying a mirrored repository in any way that it no longer matches the external service's configuration of which repositories to mirror (e.g via `gitlab.projectQuery`, `github.repositoryQuery`, `github.repos`, etc) **SHOULD** result in newly accessible repositories to be mirrored and no longer accessible repositories to be soft-deleted.
1. **Modifying configuration**: Modifying an external service's configuration of which repositories to mirror (e.g via `gitlab.projectQuery`, `github.repositoryQuery`, `github.repos`, etc) in any way that it results in some previously mirrored repositories no longer being selected to be mirrored **SHOULD** result in those repositories to be soft-deleted and previously unselected repositories now being selected, to be mirrored.

## Dependencies

Besides the previous work done as part of ["Handle renames and deletions of mirrored repositories" (#914)](https://github.com/sourcegraph/sourcegraph/issues/914), this proposal requires additional UI changes for:

1. Showing repositories as deleted, including the reason for which they were deleted.

## Implementation

This section describes the possible implementation paths for the above proposal and some key questions that need to be answered.

### Tasks

- [ ] Update [createEnableUpdateRepos](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repos/util.go#L70:6-70:29) to operate on the full list of connections instead of a single connection so that it can compute what has been added/deleted and then handle the deleted case appropriately.
- [ ] Open a PR to make this a document in `doc/proposals` and continue discussion there
- [ ] TODO

### Questions

**How do we handle the same repository being selected from multiple external services and/or from multiple connections in a single external service?**

Multiple matches of the same repository across multiple connections of the **same** external service are de-duplicated based on that external service's stable identifier for a single repository.

We don't de-deduplicate the same repository across different external services. See https://github.com/sourcegraph/sourcegraph/issues/1467#issuecomment-447979678 for more context.

**What happens if a repository is renamed?**

Based on the repository's stable identifier as provided by its external service, we can detect that its name has changed and just update it in our database.

Cached and derived data (git-server, search index) is re-fetched and re-computed. (@nicksnyder : How can this be triggered this?)

Cached and derived data associated with the previous repository name is lazily deleted.

**What happens to a repository when no external service configuration selects it to be mirrored any longer?**

It's soft-deleted.

**What happens to data that is attached to repositories (e.g. code discussions) when a repository is deleted?**

This data is soft-deleted along with the repository and can be restored with it too.

**What happens when a token is deleted from the database?**

All repositories that can't be seen anymore because the token is gone will be soft-deleted.

**Comment by sqs** [comment]
> Renaming repositories: Renaming a mirrored repository in an external service SHOULD result in the newly named repository to be mirrored and the previous repository to be marked as disabled.

The desired behavior for renaming is that the repository’s name is also changed on Sourcegraph. The behavior described above is a good intermediate step, though. 

---


Re: “modifying a repository”, what kinds of modifications are you thinking of? Right now there is no way for a user on Sourcegraph to modify the repository—all metadata comes from the external service. I think we should just assume that we’ll always prevent users from manually editing metadata of repositories from external services.

**Comment by tsenart** [comment]
We can implement that desired behavior for renaming with de-duplication. Not sure how otherwise.

As for the modifications, I was referring to upstream repo metadata modifications. For instance, if `github.repositoryQuery` filters repos with certain labels, and someone adds that label to a repo at a later point, we should start mirroring that repo.

Sent via Superhuman iOS ( https://sprh.mn/?vip=tomas@sourcegraph.com )

**Comment by nicksnyder** [comment]
Thanks for writing this up!

> 1. Removed permissions: Removing permissions from an access token used to read previously mirrored repositories from an external service SHOULD result in those now inaccessible repositories to be clearly marked as disabled in the Sourcegraph UI with the reason for which they are disabled made clear to the user.

I have two thoughts about this (and that also apply to 3, 4, 5, 6).

(1)
Having a "disabled" state as you describe doesn't seem correct to me from first principles. Are you aware of any code hosts that behave like this? We aren't a code host, but we are similar and I don't see an argument why our users would expect any sort of grace period. As a user, I would be confused/annoyed by any duplication on my Sourcegraph instance. I just want my Sourcegraph instance to instantly mirror my code host (since my code host is the source of truth).

(2)
Our current system already has two states:
- **enabled** contents are searchable and kept up to date
- **disabled** repo is part of list of available repos synced from code host, but is not searchable or kept up to date

In your proposal it is unclear if you are proposing to change the definition of "disabled" or if you are proposing adding a third state.

In either case, it sound like a decent amount of added complexity (both in terms of implementation, but also in terms of a user understanding how our product works), which, all else equal, I would like to avoid.

> Meta-data comparison is unreliable. 

I assume that if a git repository is renamed in most code hosts, there exists a stable ID that does not change. I think we should plan to use that, instead of comparing git history as long as you can verify this is true for the code hosts that we support. No migration work of our own data will need to be done on renames because the ID has not changed. If there exist code hosts now or in the future that don't have a stable ID, we could maybe find a way generate a pseudo id from the git history (e.g. commit hash of the first head ref in the repo).

**Comment by tsenart** [comment]
> Having a "disabled" state as you describe doesn't seem correct to me from first principles. Are you aware of any code hosts that behave like this? We aren't a code host, but we are similar and I don't see an argument why our users would expect any sort of grace period. As a user, I would be confused/annoyed by any duplication on my Sourcegraph instance. I just want my Sourcegraph instance to instantly mirror my code host (since my code host is the source of truth).

I am not aware of any code-hosts the have a grace period. But I know that code-hosts implement fail-safe UX patterns to prevent mistakes with sensitive or destructive operations. For instance, Github asks you to enter the name of a repository before deleting it in order to reduce the likelihood of absent mindedness ending up in unintended consequences.

Sourcegraph not being a code host is a critical difference in my mind. We are working with mirrored repositories from the sources of truth. We compute secondary indexes like the search index (which is costly), capture associated data like code discussions, and, in the future, I'm assuming other work will be done that it'd be wasteful, or even wrong, to throw away.

Since repository mirroring happens in the background, we cannot synchronously ask an admin to confirm that the {deletion|rename|etc} was actually intended. Having a grace period allows for recovery of mistakes.

Now, I think I didn't make it clear that repositories in this `disabled` state would not actually appear in search results, or any other user facing page, so, from a user perspective, they are deleted. It's just that the admin can undo whatever mistaken upstream code host operation during the grace period, and everything would resume working as before, without loosing any data (like code discussions, or re-computing expensive indexes.

As for UX examples of this pattern in different contexts, there are many. For instance:

- Your desktop trash bin (e.g in macOS): You add things to the trash but can recover them if you did it by mistake before deleting them permanently.
- iOS Photos: You delete a photo and it ends up in a deleted album, which gets permanently deleted after a certain time period, unless you explicitly do it. Again, so that you can recover.

Despite the above, we can and should give the option to the very confident admin to immediately delete things. This would be done with setting the grace period to `0`.

Newly discovered / selected repositories are not bound to a grace period and are immediately mirrored and processed.

> Our current system already has two states:
> enabled contents are searchable and kept up to date disabled repo is part of list of available repos synced from code host, but is not searchable or kept up to date
> In your proposal it is unclear if you are proposing to change the definition of "disabled" or if you are proposing adding a third state.

Aha! Some discovery happening right now :-) OK, so, in which ways can repositories end up in a disabled state? I wasn't aware of it.

In any case, we can replace all occurrences of the above `disabled` with a third state named `deleted` that has the grace period semantics described before AND the `disabled` semantics (not searchable, etc).

> I assume that if a git repository is renamed in most code hosts, there exists a stable ID that does not change. I think we should plan to use that, instead of comparing git history as long as you can verify this is true for the code hosts that we support. No migration work of our own data will need to be done on renames because the ID has not changed. If there exist code hosts now or in the future that don't have a stable ID, we could maybe find a way generate a pseudo id from the git history (e.g. commit hash of the first head ref in the repo).

Good call! I didn't think of this and it may be good enough for now. However, it won't work for de-duplicating repositories **across** external services.

Here's a somehow realistic example of a setup where this could matter: A company is using both Github.com and Github Enterprise. They want to use Sourcegraph with both of them. For contrived reasons, they have some repositories mirrored from their Github.com organization to their Github Enterprise instance.

By relying on a stable identifier alone, we can't de-duplicate these mirrored repos across different external services. We have to rely on Git history.

Using Git history is a more general solution to the problem, but it'd be more work to implement. What's our stance on this edge-case? Do we care about it right now? Can we foresee any other advantages or disadvantages of doing it via git history now versus later?

**Comment by tsenart** [comment]
Another thing that occurred to me, in addition to my above proposal: We should e-mail admins a daily digest of the changes in connected / mirrored repositories. This would give them the chance to react to unintended changes or to accelerate intended ones (bypassing the grace period).

**Comment by nicksnyder** [comment]
> Github asks you to enter the name of a repository before deleting it in order to reduce the likelihood of absent mindedness ending up in unintended consequences.

> Since repository mirroring happens in the background, we cannot synchronously ask an admin to confirm that the {deletion|rename|etc} was actually intended. Having a grace period allows for recovery of mistakes.

In my mind, the fact that GitHub makes you double confirm is a reason why it is ok for us to immediately delete (the user already confirmed their intentions), not a reason why we should do the same.

> Sourcegraph not being a code host is a critical difference in my mind. We are working with mirrored repositories from the sources of truth. We compute secondary indexes like the search index (which is costly), capture associated data like code discussions, and, in the future, I'm assuming other work will be done that it'd be wasteful, or even wrong, to throw away.

Anything that is just a cache (search index, git server), is totally fine to throw away. Data on gitserver is already lazily (not immediately) removed periodically. I don't know how we currently deal with cleaning up search index data (something for you to figure out as part of this project perhaps). It is not going to be surprising to users that if they accidentally delete their monorepository from their code host, that there will be downstream repercussions to Sourcegraph (e.g. it might need to re-clone if we have already lazily deleted the data).

For the data that is not just a cache (e.g. code discussions), that data should merely be [soft deleted](https://github.com/sourcegraph/sourcegraph/blob/master/doc/dev/postgresql.md#hard-vs-soft-deletes). I then rather spend effort making this flow work:
1. User accidentally deletes repo from code host
2. Sourcegraph soft deletes all associated data
3. User realizes mistake and un-deletes that repo from code host
4. When Sourcegraph syncs, it detects the existence of soft-deleted data so instead of inserting a new repo row in the db, it just un-soft deletes the existing row. All the other data will still be attached.

This sounds much simpler to me to implement and explain.

>  It's just that the admin can undo whatever mistaken upstream code host operation during the grace period, and everything would resume working as before, without loosing any data (like code discussions, or re-computing expensive indexes.

Yes, as I proposed above, the admin shouldn't lose any data, but this doesn't require a grace period on Sourcegraph. They made a mistake on their code host so the only action they should need to take is to fix that mistake on their code host and then Sourcegraph should intelligently restore the data.

> OK, so, in which ways can repositories end up in a disabled state? I wasn't aware of it.

This is generally managed by the site admin (e.g. https://sourcegraph.sgdev.org/site-admin/repositories). By default all repositories are in the disabled state. Site admins either manually enable/disable repositories one by one, script enablement/disablement using our GraphQL API, or configure [initialRepositoryEnablement](https://docs.sourcegraph.com/admin/site_config/all#initialrepositoryenablement-boolean) to `true`.

> In any case, we can replace all occurrences of the above `disabled` with a third state named `deleted` that has the grace period semantics described before AND the `disabled` semantics (not searchable, etc).

In either your proposal or my proposal, I would recommend the way to do this is to add a `deleted_at` column to the `repo` table and treat this as a soft delete.

> Good call! I didn't think of this and it may be good enough for now. However, it won't work for de-duplicating repositories across external services.

Very generally I think deduplicating across services is out of scope.

> Here's a somehow realistic example of a setup where this could matter: A company is using both Github.com and Github Enterprise. They want to use Sourcegraph with both of them. For contrived reasons, they have some repositories mirrored from their Github.com organization to their Github Enterprise instance.

I don't understand this use case enough to be able to make design decisions based on it. I would expect that if a company has mirrored something to their GHE instance, then that should be a source of truth for that repo and they should configure their GitHub.com connection (using access tokens or repo query) to not sync repos that they have on their GHE instance.

> Using Git history is a more general solution to the problem, but it'd be more work to implement. What's our stance on this edge-case? Do we care about it right now? Can we foresee any other advantages or disadvantages of doing it via git history now versus later?

It isn't clear to me that it is possible to avoid false positives (e.g. identifying something as a "copy" which the customer would prefer to treat as a distinct fork).

To cut scope and complexity in the absences of a clear customer need, I think it shouldn't rely on git history for this.

**Comment by tsenart** [comment]
> In my mind, the fact that GitHub makes you double confirm is a reason why it is ok for us to immediately delete (the user already confirmed their intentions), not a reason why we should do the same.

Indeed, I agree. But there are other actions that admins (or non-admins) can perform upstream that will change the resulting set of repositories to be mirrored and processed on Sourcegraph: renames, token scope / permissions change, anything else affecting the results of the search query (if used) (e.g. `github.repositoryQuery`).

Do all of these actions expose the necessary UX confirmations that we feel are necessary, given that downstream, in Sourcegraph, significant changes are bound to them?

> Anything that is just a cache (search index, git server), is totally fine to throw away.

This is non-obvious to me. Would you please explain what makes it OK to throw away indexes? I know we can recompute things. Is this computation so cheap and fast that it's a non-issue? Will it remain this way in the foreseeable future?

> Data on gitserver is already lazily (not immediately) removed periodically.

Where can I learn more about the retention policy / logic?

> They made a mistake on their code host so the only action they should need to take is to fix that mistake on their code host and then Sourcegraph should intelligently restore the data.

Alright, it seems that there's agreement that restoring mistakenly deleted "data" is an important use-case we should cover. Just to verify that I understand the difference between what you're suggesting and what I proposed, please confirm that this is also your understanding:

Your proposal:

- **Eagerly** fully-delete any secondary indexes or caches: search index, git-server data, etc.
- **Eagerly** soft-delete database records with `deleted_at` columns: repos and associated data (e.g. code discussions).
- **Restore** by clearing out `deleted_at` columns and re-computing secondary indexes and re-filling caches.

My original proposal:

- **Lazily** delete any secondary indexes, caches and database records after a GC period / grace period.
- **Eagerly** mark all things to be deleted (gitserver data, search indexes, database records) with future-time `deleted_at` timestamp, somewhere (in Postgres for database records).
- **Restore** within grace period by clearing out `deleted_at` markers. After grace period users can't restore. Everything would be done from scratch, and associated data would be gone.

> It isn't clear to me that it is possible to avoid false positives (e.g. identifying something as a "copy" which the customer would prefer to treat as a distinct fork).
> To cut scope and complexity in the absences of a clear customer need, I think it shouldn't rely on git history for this.

Makes sense. Sounds good to me. 👍

**Comment by tsenart** [comment]
> As long as a user can add back a repo to Sourcegraph and have all previous data restored (as I proposed), then I don't see a reason why we would need to give them a grace period.

For posterity, the motivation for me to propose the deletion grace period was the wrong assumption that recomputing secondary indexes (search index) and re-filling caches (git-server) was something we wanted to avoid because it was expensive (particularly in the search index case). Tangentially, I also wasn't aware that both search and git-server had their own eviction policy implemented.

Having learned that, your proposal make total sense to me. After thinking through some details, I'll edit the issue description accordingly, followed by writing a step-by-step breakdown of implementation steps which I'd like you to review before I start working through it.

**Comment by tsenart** [comment]
Here's the patch of the proposal, for transparency.

```patch
9c9
< 1. **Removed permissions**: Removing permissions from an access token used to read previously mirrored repositories from an external service **SHOULD** result in those now inaccessible repositories to be clearly marked as _disabled_ in the Sourcegraph UI with the reason for which they are disabled made clear to the user.
---
> 1. **Removed permissions**: Removing permissions from an access token used to read previously mirrored repositories from an external service **SHOULD** result in those now inaccessible repositories to be soft-deleted from Sourcegraph, with the reason for which they were deleted made clear to the user.
11,14c11,14
< 1. **Renaming repositories**: Renaming a mirrored repository in an external service **SHOULD** result in the newly named repository to be mirrored and the previous repository to be marked as _disabled_.
< 1. **Deleting repositories**: Deleting a mirrored repository in an external service **SHOULD** result in the repository being marked as _disabled_. Disabled repositories have a default (but configurable) grace period of `1 week`, after which they will be permanently deleted. If a deleted repository is restored in the external service during its deletion grace period, Sourcegraph automatically cancels its deletion and resumes mirroring.
< 1. **Modifying repositories**: Modifying a mirrored repository in any way that it no longer matches the external service's configuration of which repositories to mirror (e.g via `gitlab.projectQuery`, `github.repositoryQuery`, `github.repos`, etc) **SHOULD** result in newly accessible repositories to be mirrored and no longer accessible repositories to be marked as _disabled_.
< 1. **Modifying configuration**: Modifying an external service's configuration of which repositories to mirror (e.g via `gitlab.projectQuery`, `github.repositoryQuery`, `github.repos`, etc) in any way that it results in some previously mirrored repositories no longer being selected to be mirrored **SHOULD** result in those repositories to be marked as _disabled_ and previously unselected repositories now being selected to be mirrored.
---
> 1. **Renaming repositories**: Renaming a mirrored repository in an external service **SHOULD** result in a graceful repository rename operation in Sourcegraph (i.e. all data is preserved, only the name changes).
> 1. **Deleting repositories**: Deleting a mirrored repository in an external service **SHOULD** result in the repository being soft-deleted via a `deleted_at` timestamp in a database column and associated application logic. Soft-deleted repositories can be restored gracefully by clearing out their `deleted_at` column. Associated data like code discussions is preserved and restorable. Cached or derived data like git-server storage and search indexes is not preserved because it's cheap to re-fetch / re-compute.
> 1. **Modifying repositories**: Modifying a mirrored repository in any way that it no longer matches the external service's configuration of which repositories to mirror (e.g via `gitlab.projectQuery`, `github.repositoryQuery`, `github.repos`, etc) **SHOULD** result in newly accessible repositories to be mirrored and no longer accessible repositories to be soft-deleted.
> 1. **Modifying configuration**: Modifying an external service's configuration of which repositories to mirror (e.g via `gitlab.projectQuery`, `github.repositoryQuery`, `github.repos`, etc) in any way that it results in some previously mirrored repositories no longer being selected to be mirrored **SHOULD** result in those repositories to be soft-deleted and previously unselected repositories now being selected, to be mirrored.
20,22c20
< 1. Showing repositories as disabled, including the reason for which they are disabled.
< 1. Showing time left in a disabled repository's grace period before it is permanently deleted.
< 1. In case we implement de-duplication (see blow), some form of showing the user all the upstream repos (remotes) that map to the local mirrored repo.
---
> 1. Showing repositories as deleted, including the reason for which they were deleted.
34c32
< Let's talk about two possible alternatives: de-duplication and no de-duplication of repositories.
---
> Multiple matches of the same repository across multiple connections of the **same** external service are de-duplicated based on that external service's stable identifier for a single repository.
36,66c34
< Without de-duplication, a single repository that's selected to be mirrored from multiple external services will be mapped to independent rows in our repos table and hence overall separate repositories across the Sourcegraph system.
< 
< _No de-duplication analysis_:
< 
< - Pro: Simple.
< - Con: Poor UX (e.g same logical repository shows up in search results multiple times)
< - Con: Inefficient storage and CPU resource utilization.
< 
< If the drawbacks of the no de-duplication approach outweigh the benefit of its implementation simplicity, then there are a few ideas we can consider on how to tackle repository de-duplication. Let's begin by considering what is required to identify two separate Git repositories as being the same.
< 
< Meta-data comparison is unreliable. We can't tell only by the name of a repository if it is the same as another. We need to go deeper and compare their contents.
< 
< A naive approach would be to recursively hash all of the files and directories of two repositories and then compare hashes. This solution is brittle because two repositories may not be in perfect sync, but still be the same repository. The hashes of the two would differ with the slightest difference in their contents, for instance if their master HEADs would differ by a single commit.
< 
< If we want to recognize two repositories as the same even if one of them is ahead in time for the same ref, we must take into account their history. This can be done using the [Lowest Common Ancestor](https://en.wikipedia.org/wiki/Lowest_common_ancestor) algorithm over a pair of Git repositories.
< 
< Using `git merge-base our-ref their-ref` works to find the latest common ancestor between two refs, even if from different repos. Different repos can be added as separate remotes with `git remote add other-repo ...`.
< 
< Once it's determined that two repositories have common histories, we can be confident that they're the same repository, even if not completely up to date with each other or if diverged at some point.
< 
< Doing this for the master branch only could be sufficient heuristically, but if necessary we do it for all branches and tags.
< 
< Armed with this knowledge, we can model a repository in the repos table to have a one to many relationship to the upstream external service repositories that it mirrors. This would map well to the concept of `git remotes`.
< 
< As for how to efficiently answer the query "Give me the repository that has the **longest** common history with this one", we'd have a few options to discuss:
< 
< - **Brute-force, no index**: Fan-out a `git merge-base` command across all repos in all `git-servers`, gather the results and find the latest common ancestor, breaking ties with commit dates.
< - **Index on git-server**: Hand-roll some sort of durable tree based index.
< - **Use [gitbase](https://github.com/src-d/gitbase)**: Leverage gitbase which seems to cover this use case and could possibly be useful for future needs.
< 
< _De-duplication analysis_: The opposite of the no de-duplication approach.
---
> We don't de-deduplicate the same repository across different external services. See https://github.com/sourcegraph/sourcegraph/issues/1467#issuecomment-447979678 for more context.
70c38,40
< Without de-duplication, we can simply `disable` the old repository and mirror the new one from scratch.
---
> Based on the repository's stable identifier as provided by its external service, we can detect that its name has changed and just update it in our database.
> 
> Cached and derived data (git-server, search index) is re-fetched and re-computed. (@nicksneyder: How can this be triggered this?)
72c42
< With de-duplication, we'd associate a new external service repo to the already existing repo (remember, one to many association) and mark the previous external repo as `disabled`.
---
> Cached and derived data associated with the previous repository name is lazily deleted.
76c46
< It's marked as `disabled` and will be deleted after the configured grace period (default `1 week`).
---
> It's soft-deleted.
80c50
< This data is deleted with the repository.
---
> This data is soft-deleted along with the repository and can be restored with it too.
84c54
< All repositories that can't be seen anymore because the token is gone will be marked as `disabled` and deleted at the end of the configured grace period.
\ No newline at end of file
---
> All repositories that can't be seen anymore because the token is gone will be soft-deleted.
```

**Comment by tsenart** [comment]
@nicksnyder: Please take a look. It's updated in the issue's description.

**Comment by nicksnyder** [comment]
> both search and git-server had their own eviction policy implemented

To be precise:
- gitserver currently only evicts on sourcegraph.com, but we might want to do this on private instances too in the new world
- I don't know what search already does off the top of my head (@keegancsmith would know but he is on vacation), but I assume we would want searcher to behave similarly as gitserver in the new world.

> Showing repositories as deleted, including the reason for which they were deleted.

Since the repository is going to be removed from the UI, it is unclear where we would show this reason. It is also unclear to me how we would distinguish different reasons (e.g. access token no longer has scope, repo query changed, etc.). I don't think we need to show a reason because the reason is always "this repo is no longer returned by the code host based on your current config".

> Cached or derived data like git-server storage and search indexes is not preserved because it's cheap to re-fetch / re-compute.

I would rephrase this to be:

"Cached or derived data like git-server storage and search indexes is preserved because it's cheap to re-fetch / re-compute."

> Cached and derived data (git-server, search index) is re-fetched and re-computed. (@nicksnyder : How can this be triggered this?)

We currently model things like this as a periodic janitor job (e.g. run every 24 hours) since we don't need to proactively purge this data.

**Comment by tsenart** [comment]
Closing this in favour of #2025.
